### PR TITLE
Feature/add rate limiting 151 redeem pallet

### PR DESCRIPTION
**Pull request by cr4pt0**
Add rate limit volume functionality and unit tests for redeem pallet.
#154

### CODE REVIEW

File: pallets/redeem/src/tests.rs [hunk 1]
```diff
@@ -1078,3 +1079,431 @@ mod spec_based_tests {
...
+				active_replace_collateral: 0,
+				banned_until: None,
+				secure_collateral_threshold: None,
+				status: VaultStatus::Active(true),
+				liquidated_collateral: 0,
+			},
+		);
+
+		let redeemer = USER;
+		let amount = 90;
```
Comment (ebma) [inline pallets/redeem/src/tests.rs:1111]: Can we use `10` as the amount here? I just would prefer to use the value that is as close to the valid threshold as possible. To make it clearer we could also introduce an extra variable at the top like
```
let volume_limit: u128 = 9;
crate::Pallet::<Test>::_rate_limit_update(
			std::option::Option::<u128>::Some(volume_limit),
			DEFAULT_COLLATERAL_CURRENCY,
			7200u64,
		);
...
...
let amount =  volume_limit + 1;
```

This way it is very clear why we choose which amount.

File: pallets/redeem/src/tests.rs [hunk 2]
```diff
@@ -1078,3 +1079,431 @@ mod spec_based_tests {
...
+		let redeem_fee = 5;
```
Comment (ebma) [inline pallets/redeem/src/tests.rs:1173]: Similar here. We can introduce a variable `volume_limit` and re-use that to set the `amount` equal to it.

File: pallets/redeem/src/tests.rs [hunk 3]
```diff
@@ -1078,3 +1079,431 @@ mod spec_based_tests {
...
+			VAULT
+		));
+	})
+}
+
+#[test]
+fn test_execute_redeem_succeeds_before_exceed_rate_limit() {
+	// POSTCONDITION: `redeemTokens` MUST be called, supplying `redeemRequest.vault`,
+	// `redeemRequest.amountBtc + redeemRequest.transferFeeBtc`, `redeemRequest.premium` and
+	// `redeemRequest.redeemer` as arguments.
```
Comment (ebma) [inline pallets/redeem/src/tests.rs:1211]: I'd remove this comment as it is probably out of place.

File: pallets/redeem/src/tests.rs [hunk 4]
```diff
@@ -1078,3 +1079,431 @@ mod spec_based_tests {
...
+			RuntimeOrigin::signed(redeemer),
+			amount,
+			stellar_address,
+			VAULT
+		));
+	})
+}
+
+#[test]
+fn test_execute_redeem_succeeds_before_exceed_rate_limit() {
```
Comment (ebma) [inline pallets/redeem/src/tests.rs:1208]: What is this test actually used for? Not sure I understand what it is supposed to show

File: pallets/redeem/src/tests.rs [hunk 5]
```diff
@@ -1078,3 +1079,431 @@ mod spec_based_tests {
...
+			TestError::RedeemCompleted,
+		);
+	})
+}
+
+#[test]
+fn test_execute_redeem_fails_after_exceed_rate_limit() {
+	// POSTCONDITION: `redeemTokens` MUST be called, supplying `redeemRequest.vault`,
+	// `redeemRequest.amountBtc + redeemRequest.transferFeeBtc`, `redeemRequest.premium` and
+	// `redeemRequest.redeemer` as arguments.
```
Comment (ebma) [inline pallets/redeem/src/tests.rs:1303]: Also remove this comment

File: pallets/redeem/src/tests.rs [hunk 6]
```diff
@@ -1078,3 +1079,431 @@ mod spec_based_tests {
...
+			TestError::ExceedLimitVolumeForIssueRequest
+		);
+	})
+}
+
+#[test]
+fn test_execute_redeem_after_exceed_rate_limit_reset_interval_succeeds() {
+	// POSTCONDITION: `redeemTokens` MUST be called, supplying `redeemRequest.vault`,
+	// `redeemRequest.amountBtc + redeemRequest.transferFeeBtc`, `redeemRequest.premium` and
+	// `redeemRequest.redeemer` as arguments.
```
Comment (ebma) [inline pallets/redeem/src/tests.rs:1403]: Also remove this

File: pallets/redeem/src/tests.rs [hunk 7]
```diff
@@ -1078,3 +1079,431 @@ mod spec_based_tests {
...
+			fee: 0,
+			transfer_fee: btc_fee.amount(),
+		});
+		assert_err!(
+			Redeem::get_open_redeem_request_from_id(&H256([0u8; 32])),
+			TestError::RedeemCompleted,
+		);
+
+		let redeemer = USER;
+		let amount = 2;
```
Comment (ebma) [inline pallets/redeem/src/tests.rs:1490]: Can you change it to use an amount of just `1` here? You'd have to change l. 1406 as well of course but I'd prefer to be as close to the 'edge' case as possible, so showing that it errors with an amount of `1` is better I suppose.

**Review by ebma — CHANGES_REQUESTED** [review]
I added some nitpicky comments again. The tests look good overall, I'd just like some things to be clearer.

File: pallets/redeem/src/tests.rs (hunk 7 above)
Comment (cr4pt0) [reply pallets/redeem/src/tests.rs:1490]: with one we will not exceed the limit rate.
```
ensure!(
				new_issue_request_amount.amount().saturating_add(current_volume) <= limit_volume,
				Error::<T>::ExceedLimitVolumeForIssueRequest
			);
```

the limit is 101 and the prev request redeem was 100.
so we need 2 here.
@ebma

File: pallets/redeem/src/tests.rs (hunk 4 above)
Comment (cr4pt0) [reply pallets/redeem/src/tests.rs:1208]: we check case when we try execute redeem before we reach the limit and everything works good as expected.

File: pallets/redeem/src/tests.rs (hunk 1 above)
Comment (cr4pt0) [reply pallets/redeem/src/tests.rs:1111]: done

**Pull request merged**